In [1]:
import pandas as pd     # A pandas convention 😅
import re               # Regex library. We will use that later.
import pathlib          # To manage paths in an OOP why.
import datetime         # To manage dates.

### Importing and exploring the csv file

In [2]:
file_path = pathlib.Path("store_data_20230116.csv")
store_data = pd.read_csv(file_path)

By defaule ***"head"*** shows the first 5 rows in a table.

In [3]:
store_data.head()

Tran ID         Cat             Item  Qty  Unit Price
0    25008     legumes  black-eyed peas  2.0        8.23
1    25008     legumes        chickpeas  2.0        5.12
2    25008      fruits           apples  2.0        9.60
3    25008      fruits           grapes  2.0        3.20
4    25044  vegetables          lettuce  3.0        9.11

You can get a lot of usefull information with ***"info"*** Like the rows count, column names, column types and the presence of null values

In [4]:
store_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10038 entries, 0 to 10037
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Tran ID     10038 non-null  int64  
 1   Cat         9956 non-null   object 
 2   Item        9966 non-null   object 
 3   Qty         10023 non-null  float64
 4   Unit Price  10011 non-null  float64
dtypes: float64(2), int64(1), object(2)
memory usage: 392.2+ KB


You can check for the distinct values in a column using ***unique***

In [5]:
store_data["Item"].unique() 
# TIP:
# You can also access a column with the dot notation
# if its name is one word without space
# so store_data.Item.unique() works too 😉

array(['black-eyed peas', 'chickpeas', 'apples', 'grapes', 'lettuce',
       'lentils', 'potatos', 'bananas', 'carrots', 'broccoli', 'LETTUCE',
       nan, 'oranges', 'navy beans', 'Oranges', 'Apples', 'Carrots',
       'Potatos', 'LENTILS', 'Broccoli', 'Chickpeas', 'CARROTS',
       'NAVY BEANS', 'Navy beans', 'BANANAS', 'BROCCOLI', 'Bananas',
       'Lentils', 'Lettuce', 'ORANGES', 'Black-eyed peas', 'Grapes',
       'GRAPES', 'CHICKPEAS', 'APPLES', 'BLACK-EYED PEAS'], dtype=object)

In [6]:
store_data.Cat.unique()

array(['legumes', 'fruits', 'vegetables', nan, 'VEGETABLES', 'Fruits',
       'Legumes', 'Vegetables', 'FRUITS', 'LEGUMES'], dtype=object)

You can also get the distinct values and their count with ***value_counts***

In [7]:
store_data.Cat.value_counts()

fruits        3381
legumes       3271
vegetables    3239
Vegetables      14
Legumes         12
FRUITS          12
VEGETABLES       9
Fruits           9
LEGUMES          9
Name: Cat, dtype: int64

A good way to explore your data across multiple columns is to use pandas.crosstab.

In [8]:
crosstab = pd.crosstab(                # By default crosstab creates a simple frequency (count) table.
                index=store_data.Item,
                columns=store_data.Cat
            )
crosstab

Cat              FRUITS  Fruits  LEGUMES  Legumes  VEGETABLES  Vegetables  \
Item                                                                        
Apples                2       0        0        0           0           0   
BANANAS               0       1        0        0           0           0   
Bananas               1       0        0        0           0           0   
Black-eyed peas       0       0        1        0           0           0   
Broccoli              0       0        0        0           1           0   
CARROTS               0       0        0        0           1           0   
CHICKPEAS             0       0        0        0           0           0   
Carrots               0       0        0        0           0           1   
Chickpeas             0       0        1        1           0           0   
GRAPES                1       2        0        0           0           0   
Grapes                0       0        0        0           0           0   
LENTILS               0       0        0        1           0           0   
LETTUCE               0       0        0        0           0           1   
Lentils               0       0        0        0           0           0   
Lettuce               0       0        0        0           0           3   
NAVY BEANS            0       0        0        0           0           0   
Navy beans            0       0        2        0           0           0   
ORANGES               0       1        0        0           0           0   
Oranges               1       1        0        0           0           0   
Potatos               0       0        0        0           1           1   
apples                0       0        0        0           0           0   
bananas               1       2        0        0           0           0   
black-eyed peas       0       0        0        1           0           0   
broccoli              0       0        0        0           0           1   
carrots               0       0        0        0           0           1   
chickpeas             0       0        0        1           0           0   
grapes                0       0        0        0           0           0   
lentils               0       0        1        2           0           0   
lettuce               0       0        0        0           0           1   
navy beans            0       0        0        0           0           0   
oranges               0       0        0        0           0           0   
potatos               0       0        0        0           2           1   

Cat              fruits  legumes  vegetables  
Item                                          
Apples                0        0           0  
BANANAS               1        0           0  
Bananas               1        0           0  
Black-eyed peas       0        1           0  
Broccoli              0        0           2  
CARROTS               0        0           2  
CHICKPEAS             0        2           0  
Carrots               0        0           2  
Chickpeas             0        0           0  
GRAPES                2        0           0  
Grapes                1        0           0  
LENTILS               0        0           0  
LETTUCE               0        0           1  
Lentils               0        1           0  
Lettuce               0        0           0  
NAVY BEANS            0        1           0  
Navy beans            0        1           0  
ORANGES               0        0           0  
Oranges               1        0           0  
Potatos               0        0           1  
apples              910        0           0  
bananas             790        0           0  
black-eyed peas       0      851           0  
broccoli              0        0         816  
carrots               0        0         809  
chickpeas             0      789           0  
grapes              810        0           0  
lentils           

To narrow it down a little...

In [9]:
crosstab.loc[
    crosstab.index.str.lower() == 'carrots',                            # Index boolean mask
    [col for col in crosstab.columns if col.lower() == 'vegetables']    # Selected columns
]

Cat      VEGETABLES  Vegetables  vegetables
Item                                       
CARROTS           1           0           2
Carrots           0           1           2
carrots           0           1         809

You can check for null values with the help of ***isna*** or ***isnull*** either for the entire dataframe, a subset of columns or a single column.

In [10]:
store_data.isnull().sum()

Tran ID        0
Cat           82
Item          72
Qty           15
Unit Price    27
dtype: int64

And to check for duplicates, we can use ***duplicated***

In [11]:
store_data.duplicated().sum()   # Duplicated records count.

175

In [12]:
store_data[store_data.duplicated(keep=False)].head(10)  # Explore the duplicated rows.

Tran ID         Cat             Item  Qty  Unit Price
75     25248     legumes  black-eyed peas  1.0        8.23
77     25248     legumes  black-eyed peas  1.0        8.23
127    25433  vegetables          potatos  1.0        1.97
128    25433  vegetables          potatos  1.0        1.97
132    25446  vegetables         broccoli  1.0        2.18
133    25446  vegetables         broccoli  1.0        2.18
150    25499  vegetables         broccoli  3.0        2.18
151    25499  vegetables         broccoli  3.0        2.18
224    25686  vegetables          potatos  4.0        1.97
225    25686  vegetables          potatos  4.0        1.97

### Data Wrangling, the fun stuff! 😃

Renaming the columns to be more convenient. Here we will remove the spaces in the column names, convert them to lower case and giving some a more meaningful name.

In [13]:
# Doing a bulk rename
store_data.columns = [col.replace(" ", "_").lower() for col in store_data.columns]

# Doing individual rename
store_data.rename(
    columns={
        "cat": "category",
        "qty": "quantity"
    },
    inplace=True    # Without this, the rename method with return a copy of the modified dataframe.
)

store_data.columns

Index(['tran_id', 'category', 'item', 'quantity', 'unit_price'], dtype='object')

We saw 175 duplicated rows during our exploration! let's get rid of those as they are likely a mistake.

In [14]:
# One way to drop the duplicates is to not select them.
count_before = len(store_data)
store_data = store_data[~store_data.duplicated()]
count_after = len(store_data)

print(f"Rows count before removing duplicates {count_before}, and after {count_after}. Thid diff is {count_before - count_after}")

Rows count before removing duplicates 10038, and after 9863. Thid diff is 175


The requirements don't need the ***tran_id*** (transaction ID) column. Let's drop it!

In [15]:
store_data.drop(
    columns="tran_id",
    inplace=True        # Again without this, a copy of the modified dataframe will be returned.
    )

store_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9863 entries, 0 to 10037
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   category    9781 non-null   object 
 1   item        9791 non-null   object 
 2   quantity    9848 non-null   float64
 3   unit_price  9836 non-null   float64
dtypes: float64(2), object(2)
memory usage: 385.3+ KB


OK, let's unify the ***item*** and ***category*** columns. We will change everything to lower case.

In [16]:
store_data.category = store_data.category.str.lower()
store_data.item = store_data.item.str.lower()

Now, let's deal with null values! Yikes 😣!
<br>As a reminder, let's check for them again.

In [17]:
store_data.isna().sum() # We used isna instead of isnull this time. Same outcome!

category      82
item          72
quantity      15
unit_price    27
dtype: int64

First, the ***quantity*** column.

In [18]:
store_data.dropna(
    subset="quantity",
    inplace=True        # You know why using this by now! I won't highlight it again.
    )

store_data.isna().sum()

category      78
item          65
quantity       0
unit_price    26
dtype: int64

We will drop the row if both the ***category*** and the ***item*** are nulls

In [19]:
store_data.dropna(
    subset=["quantity", "item"],
    inplace=True
)

store_data.isna().sum() # Great! now the item column in null-free 😊

category      48
item           0
quantity       0
unit_price    15
dtype: int64

Next, the ***category*** column.

In [20]:
categories = store_data[["category", "item"]].groupby(by=["category", "item"], as_index=False).count()
categories

category             item
0       fruits           apples
1       fruits          bananas
2       fruits           grapes
3       fruits          oranges
4      legumes  black-eyed peas
5      legumes        chickpeas
6      legumes          lentils
7      legumes       navy beans
8   vegetables         broccoli
9   vegetables          carrots
10  vegetables          lettuce
11  vegetables          potatos

In [21]:
categories_dict = {row["item"]: row["category"] for _, row in categories.iterrows()}
categories_dict

{'apples': 'fruits',
 'bananas': 'fruits',
 'grapes': 'fruits',
 'oranges': 'fruits',
 'black-eyed peas': 'legumes',
 'chickpeas': 'legumes',
 'lentils': 'legumes',
 'navy beans': 'legumes',
 'broccoli': 'vegetables',
 'carrots': 'vegetables',
 'lettuce': 'vegetables',
 'potatos': 'vegetables'}

In [22]:
store_data.category = store_data.item.apply(lambda x: categories_dict[x])
# TIP:
# You pass pass custom functions to a dataframe or a series (column) with the apply() method.
# It can be a Lambda function as in this line of code.
# The function's parameters' count must be the same as the passed elements.
# Here we only pass values from one column, so only one parameter is used in the lambda function.

store_data.isna().sum()

category       0
item           0
quantity       0
unit_price    15
dtype: int64

And for the last remaining null-containing column ***unit_price***, we can do like we did for the ***quantity*** column assuming unified unit price across items.

In [23]:
unit_prices = store_data[["item", "unit_price"]].groupby(by=["item", "unit_price"], as_index=False).count()
unit_prices_dict = {row["item"]: row["unit_price"] for _, row in unit_prices.iterrows()}
unit_prices_dict

{'apples': 9.6,
 'bananas': 6.72,
 'black-eyed peas': 8.23,
 'broccoli': 2.18,
 'carrots': 6.05,
 'chickpeas': 5.12,
 'grapes': 3.2,
 'lentils': 1.73,
 'lettuce': 9.11,
 'navy beans': 0.24,
 'oranges': 9.44,
 'potatos': 1.97}

In [24]:
store_data.unit_price = store_data.item.apply(lambda x: unit_prices_dict[x])

store_data.isna().sum()

category      0
item          0
quantity      0
unit_price    0
dtype: int64

Yaaaay 😀! No null values!

Now that there aren't any duplicates nor null values in our data, Let's enrich it for our analysis.

First, we will change the ***quantity*** column's type to int. Quantity can't be a fraction in our example 🤷‍♂️

In [25]:
store_data.quantity = store_data.quantity.astype(int)

Then, we will add the ***total_amount*** column as ***quantity*** x ***unit_price***.

In [26]:
store_data["total_amount"] = store_data.quantity * store_data.unit_price

Next, will add a ***done_on*** date column. For future considerations if more that one files are analysed for more than one day, it will hold the date when the file was created as denoted in the file name. (remember the re module that we imported? now we will use it 😊).

In [27]:
file_date = re.search(r"\d+", file_path.stem).group()       # Regex to match all the digits in the file name without the extension (stem)
file_date = datetime.datetime.strptime(file_date, "%Y%m%d") # Parse a date object from the previouse match.

store_data["done_on"] = file_date

Now to have a look at what our data looks like

In [28]:
store_data.head()

category             item  quantity  unit_price  total_amount    done_on
0     legumes  black-eyed peas         2        8.23         16.46 2023-01-16
1     legumes        chickpeas         2        5.12         10.24 2023-01-16
2      fruits           apples         2        9.60         19.20 2023-01-16
3      fruits           grapes         2        3.20          6.40 2023-01-16
4  vegetables          lettuce         3        9.11         27.33 2023-01-16

And save the new data to file for future use.

In [29]:
processed_file_path = pathlib.Path(file_path.stem + "_process.csv")
store_data.to_csv(
    processed_file_path, 
    index=False             # To skip the index that pandas added from the new file.
    )

### And now, the analysis 🤩

First, the total revenue for each item category

In [30]:
(
    store_data[["category", "total_amount"]]
    .groupby(by="category")
    .sum()
)

total_amount
category                
fruits          73964.48
legumes         38463.31
vegetables      46240.37

Next, the top 10 purchased items

In [31]:
(
    store_data[["item", "quantity"]]
    .groupby(by="item")
    .sum()
    .sort_values(by="quantity", ascending=False)
    .head(3)
)

quantity
item                     
apples               2724
oranges              2602
black-eyed peas      2567

# And that's all folks! See you in the next notebook 😃